### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by floppyzedolfin**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,41 @@
+PROJECT=argo
+VERSION=1.0.0
+# CGO_ENABLED=0 means we're not looking for C libs when using the network packages (these libs are absent in alpine images)
```
Comment (alienorlatour) [inline Makefile]: Why are you talking about Alpine, when both services are built from scratch?

File: go.mod [hunk 2]
```diff
@@ -0,0 +1,11 @@
+module github.com/floppyzedolfin/argo
+
+go 1.15
```
Comment (alienorlatour) [inline go.mod:3]: ha!

File: pkg/logger/logger.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
+package logger
```
Comment (alienorlatour) [inline pkg/logger/logger.go:1]: This could live in another repo of yours. Some day.

File: services/front/service/cmd/main.go [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package main
+
+import (
+	"github.com/floppyzedolfin/argo/pkg/logger"
+	"github.com/floppyzedolfin/argo/services/front/service/internal"
+)
+
+const frontPort = 8411
```
Comment (alienorlatour) [inline services/front/service/cmd/main.go:8]: This could easily be a parameter. Another time.

File: pkg/logger/logger.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"runtime"
+	"strings"
+	"time"
+)
+
+type level int
```
Comment (alienorlatour) [inline pkg/logger/logger.go:13]: This is not exposed but could still use a comment.

File: pkg/logger/logger_test.go [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+	// Output: {"time":"2021-03-29T12:00:00Z","level":"Info","message":"Hello World!","caller":"logger.ExampleLogPrintResults"}
+}
+
+func ExampleLogNoOutput() {
+	// reset default values for the purpose of  testing Log
+	defaultLevel = Warning
+	clock = brokenClock
+
+	Log(Info, "Hello %s", "World!")
+	// Output:
```
Comment (alienorlatour) [inline pkg/logger/logger_test.go:22]: :O

File: services/front/service/internal/endpoints/load/load.go [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+	"github.com/floppyzedolfin/argo/pkg/logger"
+	"github.com/floppyzedolfin/argo/services/portdomain/client"
+)
+
+// Load calls the backend for each entry in the input file
+func Load(ctx *fiber.Ctx, req Request) (Response, *fiber.Error) {
+	logger.Log(logger.Info, "in load endpoint")
+	f, err := os.Open(req.Input)
+	if err != nil {
+		return Response{}, fiber.NewError(fiber.StatusUnauthorized, fmt.Sprintf("unable to open file '%s': %s", req.Input, err.Error()))
```
Comment (alienorlatour) [inline services/front/service/internal/endpoints/load/load.go:18]:
```suggestion
		return Response{}, fiber.NewError(fiber.StatusUnauthorized, fmt.Sprintf("unable to open file '%s': %w", req.Input, err.Error()))
```

File: services/front/service/internal/endpoints/load/upserter.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+
+	"github.com/floppyzedolfin/argo/services/portdomain/client/portdomain"
+)
+
+// upserter is a utility that knows how to talk with the backend
```
Comment (alienorlatour) [inline services/front/service/internal/endpoints/load/upserter.go]: Well no it doesn't. It knows how to apply a func, which may know how to talk with the backend.

File: services/front/service/internal/handlers/load.go [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+	"github.com/floppyzedolfin/argo/pkg/logger"
+	"github.com/floppyzedolfin/argo/services/front/service/internal/endpoints/load"
+)
+
+// Load wraps the front service's load endpoint, and is in charge of decoding the request and setting the status
+func Load(ctx *fiber.Ctx) error {
+	req := new(load.Request)
+	if err := ctx.BodyParser(req); err != nil {
+		return ctx.Status(fiber.StatusBadRequest).JSON(fiber.Map{"error": fmt.Sprintf("unable to parse body as a request: %s", err.Error())})
```
Comment (alienorlatour) [inline services/front/service/internal/handlers/load.go:16]:
```suggestion
		return ctx.Status(fiber.StatusBadRequest).JSON(fiber.Map{"error": fmt.Sprintf("unable to parse body as a request: %w", err.Error())})
```

File: services/portdomain/client/README.md [hunk 10]
```diff
@@ -0,0 +1,7 @@
+# README
```
Comment (alienorlatour) [inline services/portdomain/client/README.md:1]: OK. I'm reading you. I like it. But this title is not totally useful.

File: services/portdomain/client/portdomain/portdomain.proto [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+message GetRequest {
+}
+
+message GetResponse {
+    repeated Port ports = 1;
+}
+
+message Port {
```
Comment (alienorlatour) [inline services/portdomain/client/portdomain/portdomain.proto:29]: Comment exposed stuff?

File: services/portdomain/service/internal/db/db.go [hunk 12]
```diff
@@ -0,0 +1,31 @@
+package db
+
+import (
+	"github.com/floppyzedolfin/argo/services/portdomain/client/portdomain"
+)
+
+// Database exposes the API we want
+type Database interface {
```
Comment (alienorlatour) [inline services/portdomain/service/internal/db/db.go]: This does not need to be defined here. Each endpoint should define a one-func interface that the struct below happens to implement.

File: services/portdomain/service/internal/db/db.go [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+	Upsert(portdomain.Port)
+	Get() map[string]portdomain.Port
+}
+
+// database is the inner structure of our database. If we want to move to something else - mongo, ... - we can change the implemetation here.
```
Comment (alienorlatour) [inline services/portdomain/service/internal/db/db.go]: Maybe you can add a babble about which sort you would pick. NoSQL?

File: services/front/service/internal/endpoints/load/testdata/invalid.json [hunk 14]
```diff
@@ -0,0 +1,2 @@
+lorem ipsum
```
Comment (alienorlatour) [inline services/front/service/internal/endpoints/load/testdata/invalid.json:1]: You can be more creative.

http://www.cupcakeipsum.com/
https://codepen.io/jacob4/full/EVqeWM

File: services/front/service/internal/endpoints/load/load.go (hunk 7 above)
Comment (floppyzedolfin) [reply services/front/service/internal/endpoints/load/load.go:18]: No, I'm `Sprintf`-ing

File: services/front/service/internal/handlers/load.go (hunk 9 above)
Comment (floppyzedolfin) [reply services/front/service/internal/handlers/load.go:16]: Sprintf can't use %w

**Pull request merged**